simtools/data_model/metadata_collector.py

```diff
@@ -34,33 +45,118 @@ def __init__(self, args_dict):
         self.io_handler = io_handler.IOHandler()
 
         self.args_dict = args_dict
+        self.data_model_name = data_model_name
+        self.schema_file = None
+        self.schema_dict = None
         self.top_level_meta = gen.change_dict_keys_case(
             data_dict=metadata_model.get_default_metadata_dict(), lower_case=True
         )
-        self.collect_product_meta_data()
+        self.input_metadata = self._read_input_metadata_from_file()
+        self.collect_meta_data()
 
-    def collect_product_meta_data(self):
+    def collect_meta_data(self):
         """
         Collect and verify product metadata from different sources.
 
         """
 
-        self._fill_association_meta_from_args(
-            self.top_level_meta["cta"]["context"]["associated_elements"]
-        )
+        self._fill_contact_meta(self.top_level_meta["cta"]["contact"])
         self._fill_product_meta(self.top_level_meta["cta"]["product"])
-        self._fill_top_level_meta_from_file(self.top_level_meta["cta"])
-        self._fill_association_id(self.top_level_meta["cta"]["context"]["associated_elements"])
         self._fill_activity_meta(self.top_level_meta["cta"]["activity"])
+        self._fill_process_meta(self.top_level_meta["cta"]["process"])
+        self._fill_context_from_input_meta(self.top_level_meta["cta"]["context"])
+        self._fill_associated_elements_from_args(
+            self.top_level_meta["cta"]["context"]["associated_elements"]
+        )
+
+    def get_data_model_schema_file_name(self):
+        """
+        Return data model schema file name.
+        The schema file name is taken (in this order) from the command line,
+        from the metadata file, from the data model name, or from the input
+        metadata file.
+
+        Returns
+        -------
+        str
+            Name of schema file.
+
+        """
+
+        try:
+            if self.args_dict["schema"]:
+                self._logger.debug(f"Schema file from command line: {self.args_dict['schema']}")
+                return self.args_dict["schema"]
+        except KeyError:
+            pass
+
+        try:
+            if self.top_level_meta["cta"]["product"]["data"]["model"]["url"]:
+                self._logger.debug(
+                    "Schema file from product metadata: "
+                    f"{self.top_level_meta['cta']['product']['data']['model']['url']}"
+                )
+                return self.top_level_meta["cta"]["product"]["data"]["model"]["url"]
+        except KeyError:
+            pass
+
+        if self.data_model_name:
+            self._logger.debug(f"Schema file from data model name: {self.data_model_name}")
+            return simtools.constants.SCHEMA_URL + self.data_model_name + ".schema.yml"
```

Comment:
OK, no problem to keep `simtools.constants` then, makes sense.

f-strings are significantly faster than adding strings and also for the purpose of consistency since I switched everything to f-string a while back.

Reply:
OK, changed.